### ISSUE TITLE
macOS 11 beta: Dylib linking broken

### ISSUE THREAD

**Issue by pcleavelin** [ziglang/zig#5715]
On the macOS 11 Big Sur beta, [changes](https://developer.apple.com/documentation/macos-release-notes/macos-big-sur-11-beta-release-notes) to how dynamic linking works breaks how zig searches for libraries.

On release 0.6.0, after creating the `zig0` executable it throws this error:

`Unable to check cache: zig-cache/h: file not found`
`make[2]: *** [CMakeFiles/zig_build_libstage2] Error 1`
`make[1]: *** [CMakeFiles/zig_build_libstage2.dir/all] Error 2`
`make: *** [all] Error 2`

Although master has a completely different error. Probably unrelated to the dylib change.
`: CommandLine Error: Option 'mc-relax-all' registered more than once!`
`LLVM ERROR: inconsistency in registered CommandLine options`
`make[2]: *** [CMakeFiles/zig_build_libstage2] Error 1`
`make[1]: *** [CMakeFiles/zig_build_libstage2.dir/all] Error 2`
`make: *** [all] Error 2`

On 0.6.0, removing the caching of libraries found via `_dyld_get_image_name()` in `Error get_compiler_id(Buf **result)` (compiler.cpp) allows zig to compile successfully. This seems to be due to `_dyld_get_image_name()` returning paths to libraries that were removed from the filesystem in the macOS beta.

Even after compiling, running `./zig build` will result in `Build Dependencies...lld: error: Unable to find library for -lSystem`

Obviously not a super high priority as macOS 11 won't be released for some months.

This is my first time posting an issue here, so please let me know if I have failed to disclose any helpful info.

**Comment by prime31** [comment]
Same result here for Big Sur:

```Scanning dependencies of target zig_build_libstage2
Unable to check cache: zig-cache/h: file not found
make[2]: *** [CMakeFiles/zig_build_libstage2] Error 1
make[1]: *** [CMakeFiles/zig_build_libstage2.dir/all] Error 2
make: *** [all] Error 2```

**Comment by mikdusan** [comment]
just highlighting this part of apple notes:

>New in macOS Big Sur 11 beta, the system ships with a built-in dynamic linker cache of all system-provided libraries. As part of this change, copies of dynamic libraries are no longer present on the filesystem. Code that attempts to check for dynamic library presence by looking for a file at a path or enumerating a directory will fail. Instead, check for library presence by attempting to dlopen() the path, which will correctly check for the library in the cache. (62986286)

**Comment by prime31** [comment]
Crossing in this PR which gets things working on Big Sur: https://github.com/ziglang/zig/pull/5880

**Comment by kubkon** [comment]
This should be fixed with #6932 so closing. If there are still problems, please reopen.
